### PR TITLE
[HandshakeToDC] Add conversion

### PR DESCRIPTION
**Pull request by mortbopet**
... progressive lowering of handshake!
This obviously requires a future addition of an `arith-to-comb/hw` pass to map arith stuff to comb/external hardware/pipelines, ...

### CODE REVIEW

File: lib/Conversion/HandshakeToDC/HandshakeToDC.cpp [hunk 1]
```diff
@@ -0,0 +1,612 @@
...
+          // Materialize !dc.token -> !dc.value<>
+          auto vt = resultType.dyn_cast<dc::ValueType>();
+          if (vt && vt.getInnerTypes().empty())
+            return pack(builder, inputs.front(), ValueRange{});
+
+          return inputs[0];
+        });
+
+    addSourceMaterialization(
+        [&](mlir::OpBuilder &builder, mlir::Type resultType,
```
Comment (Dinistro) [inline lib/Conversion/HandshakeToDC/HandshakeToDC.cpp:98]: NIT: Same here

File: lib/Conversion/HandshakeToDC/HandshakeToDC.cpp [hunk 2]
```diff
@@ -0,0 +1,612 @@
...
+          return inputs[0];
+        });
+  }
+};
+
+template <typename OpTy>
+class DCOpConversionPattern : public OpConversionPattern<OpTy> {
+public:
+  using OpConversionPattern<OpTy>::OpConversionPattern;
+  using OpAdaptor = typename OpConversionPattern<OpTy>::OpAdaptor;
```
Comment (Dinistro) [inline lib/Conversion/HandshakeToDC/HandshakeToDC.cpp:123]: NIT: 
```suggestion
  using OpAdaptor = typename OpTy::Adaptor;
```

File: lib/Conversion/HandshakeToDC/HandshakeToDC.cpp [hunk 3]
```diff
@@ -0,0 +1,612 @@
...
+
+  DCOpConversionPattern(MLIRContext *context, TypeConverter &typeConverter,
+                        ConvertedOps *convertedOps)
+      : OpConversionPattern<OpTy>(typeConverter, context),
+        convertedOps(convertedOps) {}
+  mutable ConvertedOps *convertedOps;
```
Comment (Dinistro) [inline lib/Conversion/HandshakeToDC/HandshakeToDC.cpp:129]: I'm not a fan of `mutable`, but I guess there is no way around it.

File: lib/Conversion/HandshakeToDC/HandshakeToDC.cpp [hunk 4]
```diff
@@ -0,0 +1,612 @@
...
+  using DCOpConversionPattern<handshake::ReturnOp>::DCOpConversionPattern;
+  using OpAdaptor =
+      typename DCOpConversionPattern<handshake::ReturnOp>::OpAdaptor;
+
+  LogicalResult
+  matchAndRewrite(handshake::ReturnOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // Locate existing output op, Append operands to output op, and move to
+    // the end of the block.
+    auto hwModule = cast<hw::HWModuleOp>(op->getParentOp());
```
Comment (Dinistro) [inline lib/Conversion/HandshakeToDC/HandshakeToDC.cpp:419]: NIT: Consider using`Operation::getParentOfType`

File: lib/Conversion/HandshakeToDC/HandshakeToDC.cpp [hunk 5]
```diff
@@ -0,0 +1,612 @@
...
+    // Convert the data-side mux to a sequence of arith.select operations.
+    // The data and control muxes are assumed one-hot and the base-case is set
+    // as the first input.
+    if (withData)
+      dataMux = inputs[0].data.front();
+
+    llvm::SmallVector<Value> controlMuxInputs = {inputs.front().token};
+    for (auto [i, input] :
+         llvm::enumerate(llvm::make_range(inputs.begin() + 1, inputs.end()))) {
+      if (withData) {
```
Comment (Dinistro) [inline lib/Conversion/HandshakeToDC/HandshakeToDC.cpp:458]: NIT: use a negated condition and continue.

File: lib/Conversion/HandshakeToDC/HandshakeToDC.cpp [hunk 6]
```diff
@@ -0,0 +1,612 @@
...
+};
+
+static hw::ModulePortInfo getModulePortInfo(TypeConverter &tc,
+                                            handshake::FuncOp funcOp) {
+  hw::ModulePortInfo ports({}, {});
+  auto *ctx = funcOp->getContext();
+  auto ft = funcOp.getFunctionType();
+
+  // Add all inputs of funcOp.
+  unsigned inIdx = 0;
```
Comment (Dinistro) [inline lib/Conversion/HandshakeToDC/HandshakeToDC.cpp:504]: This seems to be unused.

File: lib/Conversion/HandshakeToDC/HandshakeToDC.cpp [hunk 7]
```diff
@@ -0,0 +1,612 @@
...
+  for (auto [index, type] : llvm::enumerate(ft.getInputs())) {
+    ports.inputs.push_back({StringAttr::get(ctx, "in" + std::to_string(index)),
```
Comment (Dinistro) [inline lib/Conversion/HandshakeToDC/HandshakeToDC.cpp:506]: Can you use `emplace_back` instead, or does this only work when there is an implemented constructor?

File: lib/Conversion/HandshakeToDC/HandshakeToDC.cpp [hunk 8]
```diff
@@ -0,0 +1,612 @@
...
+                            hw::PortDirection::INPUT, tc.convertType(type),
+                            index, hw::InnerSymAttr{}});
+    inIdx++;
+  }
+
+  // Add all outputs of funcOp.
+  for (auto [index, type] : llvm::enumerate(ft.getResults())) {
+    ports.outputs.push_back({StringAttr::get(ctx, "in" + std::to_string(index)),
```
Comment (Dinistro) [inline lib/Conversion/HandshakeToDC/HandshakeToDC.cpp:514]: Shouldn't that be `"out"`?

File: lib/Conversion/HandshakeToDC/HandshakeToDC.cpp [hunk 9]
```diff
@@ -0,0 +1,612 @@
...
+    target.addLegalOp<mlir::ModuleOp>();
+
+    // The various patterns will insert new operations into the module to
+    // facilitate the conversion - however, these operations must be
+    // distinguishable from already converted operations (which may be of the
+    // same type as the newly inserted operations). To do this, we mark all
+    // operations which have been converted as legal, and all other operations
+    // as illegal.
+    target.markUnknownOpDynamicallyLegal(
+        [&](Operation *op) { return convertedOps.contains(op); });
```
Comment (Dinistro) [inline lib/Conversion/HandshakeToDC/HandshakeToDC.cpp:583]: Are you adding new handshake operations or why is this required?

**Review by Dinistro — COMMENTED** [review]
I did a pass over the implementation and dropped a bunch of comments. 

I plan to do a more in depth walk through, but so far it looks very nice. Thanks for pushing on this.

**Review by Dinistro — COMMENTED** [review]
I did a pass over the implementation and dropped a bunch of comments. 

I plan to do a more in depth walk through, but so far it looks very nice. Thanks for pushing on this.

**Review by Dinistro — COMMENTED** [review]
I did a pass over the implementation and dropped a bunch of comments. 

I plan to do a more in depth walk through, but so far it looks very nice. Thanks for pushing on this.

**Review by Dinistro — COMMENTED** [review]
I did a pass over the implementation and dropped a bunch of comments. 

I plan to do a more in depth walk through, but so far it looks very nice. Thanks for pushing on this.

**Review by Dinistro — COMMENTED** [review]
I did a pass over the implementation and dropped a bunch of comments. 

I plan to do a more in depth walk through, but so far it looks very nice. Thanks for pushing on this.

**Review by Dinistro — COMMENTED** [review]
I did a pass over the implementation and dropped a bunch of comments. 

I plan to do a more in depth walk through, but so far it looks very nice. Thanks for pushing on this.

**Review by Dinistro — COMMENTED** [review]
I did a pass over the implementation and dropped a bunch of comments. 

I plan to do a more in depth walk through, but so far it looks very nice. Thanks for pushing on this.

**Review by mikeurbach — APPROVED** [review]
Woohoo! This looks good to me, nice use of dialect conversion.

File: lib/Conversion/HandshakeToDC/HandshakeToDC.cpp (hunk 3 above)
Comment (mortbopet) [reply lib/Conversion/HandshakeToDC/HandshakeToDC.cpp:129]: Yea... i'm super open to other suggestions but this was the easiest way i found to mark operations as legal - either that, or add an attribute to the new ops.

File: lib/Conversion/HandshakeToDC/HandshakeToDC.cpp (hunk 9 above)
Comment (mortbopet) [reply lib/Conversion/HandshakeToDC/HandshakeToDC.cpp:583]: This is needed for all of the `arith` ops that get created as part of the handshake ops (e.g. `arith.select` for `handshake.mux`). There's a bit of a dilemma here seeing as **all** operations need to be converted/touched in a `handshake.func` - which is done so by `UnitRateConversionPattern` (when no other pattern applies). However, we obviously don't want to run that pattern on these newly created ops since they _do not_ have handshake semantics... suggestions are welcome 🙂!

Comment (Dinistro) [reply lib/Conversion/HandshakeToDC/HandshakeToDC.cpp:569]: Alright, now I understand the problem. So given the current handshake design, this is indeed necessary.

File: lib/Conversion/HandshakeToDC/HandshakeToDC.cpp (hunk 3 above)
Comment (Dinistro) [reply lib/Conversion/HandshakeToDC/HandshakeToDC.cpp:127]: It's a shame that one cannot register listeners on the rewriter to get out the set of created ops in a clean manner.

**Review by Dinistro — APPROVED** [review]
LGTM!
